rasa/core/featurizers/tracker_featurizers.py

```diff
@@ -240,6 +247,299 @@ def featurize_trackers(
+
+            tracker_state_features_chunk = self._featurize_states(
+                trackers_as_states_chunk, interpreter
+            )
+            label_ids_chunk = self._convert_labels_to_ids(
+                trackers_as_actions_chunk, domain
+            )
+            entity_tags_chunk = self._create_entity_tags(
+                trackers_as_entities_chunk, interpreter, bilou_tagging
+            )
+            data_chunk_file = self._persist_chunk(
+                data_chunk_dir,
+                f"{chunk_index}_chunk.tfrecord",
+                tracker_state_features_chunk,
+                label_ids_chunk,
+                entity_tags_chunk,
+            )
+
+            data_chunk_files.append(
+                DataChunkFile(Path(data_chunk_file), len(tracker_state_features_chunk))
+            )
+
+        return data_chunk_files
+
+    @staticmethod
+    def _to_tf_features(
+        turn_state_features: List[Dict[Text, List["Features"]]],
+        turn_label_ids: np.array,
+        turn_entity_tags: List[Dict[Text, List["Features"]]],
+    ) -> Dict[Text, tf.train.Feature]:
+        tf_features = {}
+        for i, state_features in enumerate(turn_state_features):
+            for features in state_features.values():
+                tf_features.update(chunk_utils.encode_features(features, f"{i}"))
+
+        tf_features[LABEL] = chunk_utils.int_feature(turn_label_ids)
+
+        for i, entity_tag in enumerate(turn_entity_tags):
+            for attribute, features in entity_tag.items():
+                tf_features.update(
+                    chunk_utils.encode_features(
+                        features, f"{i}{chunk_utils.TF_RECORD_KEY_SEPARATOR}{attribute}"
+                    )
+                )
+
+        return tf_features
+
+    def _persist_chunk(
+        self,
+        dir_path: Text,
+        filename: Text,
+        tracker_state_features_chunk: List[List[Dict[Text, List["Features"]]]],
+        label_ids_chunk: np.ndarray,
+        entity_tags_chunk: List[List[Dict[Text, List["Features"]]]],
+    ) -> Text:
+        """Stores the chunk as TFRecord file to disk.
+
+        Args:
+            dir_path: The path to the directory.
+            filename: The filename.
+
+        Returns:
+            The absolute file path the chunk is stored to.
+        """
+        file_path = os.path.join(dir_path, filename)
+
+        # Write to TFRecord
+        with tf.io.TFRecordWriter(file_path) as tfwriter:
+            for turn_state_features, turn_label_ids, turn_entity_tags in zip(
+                tracker_state_features_chunk, label_ids_chunk, entity_tags_chunk
+            ):
+                example = tf.train.Example(
+                    features=tf.train.Features(
+                        feature=self._to_tf_features(
+                            turn_state_features, turn_label_ids, turn_entity_tags
+                        )
+                    )
+                )
```

Comment:
Hmm, we should check if we can merge this function and `persist_chunk` inside rasa.shared.nlu.training_data.training_data.py` and create a generic TFRecord writer function that can be reused.